### PR TITLE
Add support for packed clients

### PR DESCRIPTION
**Pull request by Jettford**
- Implements pack file interface
- Automatically extracts cdclient for later conversion
- Implements pack file istream buffer so file reading can be done as normal
- Navmeshes were moved to the server root so they can now be automatically extracted

### CODE REVIEW

**Review by aronwk-aaron — APPROVED** [review]

File: dGame/dInventory/Item.cpp [hunk 1]
```diff
@@ -340,18 +341,23 @@ void Item::DisassembleModel() {
 	std::string renderAsset = result.fieldIsNull(0) ? "" : std::string(result.getStringField(0));
 	std::vector<std::string> renderAssetSplit = GeneralUtils::SplitString(renderAsset, '\\');
 
-	std::string lxfmlPath = "res/BrickModels/" + GeneralUtils::SplitString(renderAssetSplit.back(), '.')[0] + ".lxfml";
-	std::ifstream file(lxfmlPath);
+	std::string lxfmlPath = "BrickModels/" + GeneralUtils::SplitString(renderAssetSplit.back(), '.')[0] + ".lxfml";
```
Comment (EmosewaMC) [inline dGame/dInventory/Item.cpp:344]: I'd prefer for this to use .at(0) so an exception is thrown should an oobe is encountered

File: dCommon/dClient/PackIndex.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	std::ifstream m_FileStream;
+
+	uint32_t m_Version;
+
+	uint32_t m_PackPathCount;
+	std::vector<std::string> m_PackPaths;
+	uint32_t m_PackFileIndexCount;
+	std::vector<PackFileIndex> m_PackFileIndices;
+
+	std::vector<Pack*> m_Packs;
```
Comment (EmosewaMC) [inline dCommon/dClient/PackIndex.h:39]: While this only gets deleted when going out of scope of main, I would like for the elements in this vector to be deleted when PackIndex is deleted.

File: dCommon/dClient/PackIndex.cpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
+#include "PackIndex.h"
+
+
+PackIndex::PackIndex(const std::filesystem::path& filePath) {
+	m_FileStream = std::ifstream(filePath / "versions" / "primary.pki", std::ios::in | std::ios::binary);
```
Comment (EmosewaMC) [inline dCommon/dClient/PackIndex.cpp:5]: Is this stream ever closed?

File: dCommon/dClient/Pack.cpp [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+		uint32_t size;
+		fread(&size, sizeof(uint32_t), 1, file);
+		pos += 4;
+
+		char* chunk = (char*)malloc(size);
+		fread(chunk, sizeof(int8_t), size, file);
+		pos += size;
+
+		int32_t err;
+		currentReadPos += ZCompression::Decompress((uint8_t*)chunk, size, reinterpret_cast<uint8_t*>(decompressedData + currentReadPos), 1024 * 256, err);
```
Comment (EmosewaMC) [inline dCommon/dClient/Pack.cpp:109]: theres a constexpr in BrickByBrickFix.h which has 1024*256 in it - We should move that to ZCompression and give this value that name.

File: dCommon/dClient/Pack.cpp [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+	char* decompressedData = (char*)malloc(pkRecord.m_UncompressedSize);
+	uint32_t currentReadPos = 0;
+
+	while (true) {
+		if (currentReadPos >= pkRecord.m_UncompressedSize) break;
+
+		uint32_t size;
+		fread(&size, sizeof(uint32_t), 1, file);
+		pos += 4;
```
Comment (EmosewaMC) [inline dCommon/dClient/Pack.cpp:102]: Similar to the pos += 4 above, what is this one doing?

File: dCommon/dClient/Manifest.cpp [hunk 6]
```diff
@@ -0,0 +1,64 @@
+#include "Manifest.h"
+
+#include "GeneralUtils.h"
+
+Manifest::Manifest(const std::filesystem::path& path) {
+	m_FileStream = std::ifstream(path, std::ios::in);
```
Comment (EmosewaMC) [inline dCommon/dClient/Manifest.cpp]: i assume this is closed somewhere?

File: dCommon/dClient/AssetManager.h [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+	bool m_SuccessfullyLoaded;
+
+	std::filesystem::path m_Path;
+	std::filesystem::path m_RootPath;
+	std::filesystem::path m_ResPath;
+
+	eAssetBundleType m_AssetBundleType = eAssetBundleType::None;
+
+	PackIndex* m_PackIndex;
+	std::unordered_map<std::string, Manifest*> m_Manifests;
```
Comment (EmosewaMC) [inline dCommon/dClient/AssetManager.h]: As like the other header, memory management here would be great.

File: dCommon/dClient/AssetManager.h [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+	AssetMemoryBuffer GetFileAsBuffer(const char* name);
+
+private:
+	void LoadManifest(const std::string& name);
+	void LoadPackIndex();
+	void UnpackRequiredAssets();
+
+	// Modified crc algorithm (mpeg2)
+	// Reference: https://stackoverflow.com/questions/54339800/how-to-modify-crc-32-to-crc-32-mpeg-2
+	uint32_t crc32b(uint32_t base, uint8_t* message, size_t l) {
```
Comment (EmosewaMC) [inline dCommon/dClient/AssetManager.h:67]: Is there a particular reason this is defined in the header file?  Changing l to be length would be ideal just for more brevity

File: dCommon/dClient/Pack.cpp [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+	if (!std::filesystem::exists(filePath)) {
+		return;
+	}
+
+	m_FileStream = std::ifstream(filePath, std::ios::in | std::ios::binary);
+
+	m_FileStream.read(m_Version, 7); 
+
+	m_FileStream.seekg(-8, std::ios::end);
```
Comment (EmosewaMC) [inline dCommon/dClient/Pack.cpp:18]: A comment explaining why this is done would be great

**Review by EmosewaMC — COMMENTED** [review]
awesome stuff - would like to see more comment blocks above functions and a couple code blocks describing what they do so future maintainers can more easily modify this code.  Links to reference docs as well would be nice so you dont need to know where the docs are to use them.

A couple vectors and pointers are not managed, like the Game::AssetManager as a whole is never deleted, so it'd be ideal if, like the other Game variables, AssetManager also gets deleted just before leaving scope of main.

File: dCommon/dClient/AssetManager.h (hunk 8 above)
Comment (EmosewaMC) [reply dCommon/dClient/AssetManager.h]: I was just asking here if there is a reason this is defined in the header file - I wasn't really asking for it to be moved, I was just curious if there was a reason or not.

**Review by aronwk-aaron — APPROVED** [review]

**Review by EmosewaMC — APPROVED** [review]
Looks great!  very close to full server setup automation

**Pull request merged**